### PR TITLE
python: implement PEP 668

### PR DESCRIPTION
**Pull request by zimbatm**
Add a marker file to the python outputs that tells pip and other tooling
following PEP 668 that they should not be installing things system-wide.

This provides better feedback to the user and also potentially avoids issues in
single-user installs where the /nix/store is owned by the user.

For more details, see <https://peps.python.org/pep-0668/>

Here is how it currently looks like:

    $ pip install requests
    error: externally-managed-environment

    × This environment is externally managed
    ╰─> To install Python packages system-wide, try 'nix-env -iA
        python3.pkgs.xyz', where xyz is the package you are trying to
        install.

        If you wish to install a non-packaged Python package,
        create a virtual environment using python3 -m venv path/to/venv.
        Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
        sure you have python3-full installed.

    note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
    hint: See PEP 668 for the detailed specification.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dvn0** [comment]
> ╰─> To install Python packages system-wide, try 'nix-env -iA

Should `nix profile install` be mentioned here like in some other error messages?

**Comment by mweinelt** [comment]
Mixed feelings about yet another recommendation to use `nix-env`, but overall a good improvement.

**Comment by 4cq2** [comment]
Why isnt  `pip install --user` mentioned in the error text? Is it not an option with PEP 668? If not, why not?

`venv` just seems like a chainsaw, when a butter knife should do just fine.

**Comment by mweinelt** [comment]
User environments are messy and pip will not be able to correctly install packages that tie into native libraries.

**Comment by 4cq2** [comment]
> pip will not be able to correctly install packages that tie into native libraries.

this works fine for me, do you have an example package to demonstrate the problem?

**Comment by mweinelt** [comment]
numpy, scipy, torch, tensorflow, protobuf, portaudio, etc.

**Comment by 4cq2** [comment]
I just ran this command without error:

~~~
> pip install --user protobuf
Collecting protobuf
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 420.6/420.6 KB 5.3 MB/s eta 0:00:00
Installing collected packages: protobuf
Successfully installed protobuf-4.22.3
~~~

can you clarify what your concern is here?

**Comment by mweinelt** [comment]
```
/tmp ❯ pip install numpy
Collecting numpy
  Downloading numpy-1.24.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 24.1 MB/s eta 0:00:00
Installing collected packages: numpy
Successfully installed numpy-1.24.3

[notice] A new release of pip is available: 23.0.1 -> 23.1.1
[notice] To update, run: pip install --upgrade pip
/tmp ❯ python3 -c "import numpy"                                                           29s  tmp 01:02:01
Traceback (most recent call last):
  File "/tmp/venv/lib/python3.10/site-packages/numpy/core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/tmp/venv/lib/python3.10/site-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/tmp/venv/lib/python3.10/site-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ImportError: libz.so.1: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/venv/lib/python3.10/site-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/tmp/venv/lib/python3.10/site-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.10 from "/tmp/venv/bin/python3"
  * The NumPy version is: "1.24.3"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: libz.so.1: cannot open shared object file: No such file or directory
```

**Comment by 4cq2** [comment]
~~~
> pip install --user numpy
Collecting numpy
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.8/14.8 MB 27.3 MB/s eta 0:00:00
Installing collected packages: numpy
Successfully installed numpy-1.24.3
~~~

**Comment by mweinelt** [comment]
Now run `python3 -c "import numpy" `

**Comment by 4cq2** [comment]
> Now run `python3 -c "import numpy" `

the command completes with no output. checking the return code reveals success. It seems at least for some users, `pip install --user` is a valid and lightweight option, compares to heavy options like `venv`. do you have further evidence against `pip install --user`?

**Comment by mweinelt** [comment]
Are you on NixOS?

**Review by FRidh — CHANGES_REQUESTED** [review]
Thank you for looking into this. We should definitely have this file, however, I do not agree with the contents of the message.

Installing *libraries* directly like that is an unsupported workflow because when you install multiple they cannot see each other. This is also on purpose and essentially follows what we do with native libraries.

Same goes for virtual env. Yes, it went into the manual, many people like to work this way, but our expressions were never designed for it.

I'd much rather we refer to the Nixpkgs manual and don't describe anything here. It also avoids rebuilds if we do want to change the text.

[Origin thread on the topic](https://discuss.python.org/t/pep-668-marking-python-base-environments-as-externally-managed/10302/30?u=fridh).

**Comment by FRidh** [comment]
> Should `nix profile install` be mentioned here like in some other error messages?

This command is still experimental. Also, it is essentially the same as `nix-env` when it comes to *installing* libraries.

**Comment by zimbatm** [comment]
> Should nix profile install be mentioned here like in some other error messages?

Not while `nix profile` is still experimental

> Why isn't pip install --user mentioned in the error text?

`pip install --user` is also disabled by the PEP. I'm unsure why, but this is part of upstream's design. For some reason another tool called `pipx` is recommended for user installs on Debian.

> Thank you for looking into this. We should definitely have this file, however, I do not agree with the contents of the message.

:+1: I wanted to show the capability and mostly took Debian's message and adapted it. The next step is to nail that message content down.

How about something like this:

```
`pip install` doesn't work with Nix because Nix is a purely functional package manager with a unique approach to package and environment management. Nix manages packages and dependencies in an isolated manner to provide reproducible builds and environments, whereas `pip` is a package installer for Python that works with more traditional environments.

To use Python with Nix, have a look at the online documentation: <https://nixos.org/manual/nixpkgs/stable/#python>.
```

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 1]
```diff
@@ -483,6 +483,19 @@ in with passthru; stdenv.mkDerivation {
     # bytecode compilations for the same reason - we don't want bytecode generated.
     mkdir -p $out/share/gdb
     sed '/^#!/d' Tools/gdb/libpython.py > $out/share/gdb/libpython.py
+
+    # Disable system-wide pip installation. See https://peps.python.org/pep-0668/.
+    cat <<'EXTERNALLY_MANAGED' > $out/lib/${libPrefix}/EXTERNALLY-MANAGED
+    [externally-managed]
+    Error=To install Python packages system-wide, try 'nix-env -iA
```
Comment (RaitoBezarius) [inline pkgs/development/interpreters/python/cpython/default.nix:490]:
```suggestion
    Error=To install Python packages system-wide, try `nix-shell -p 'python3.withPackages (ps: [ ps.xyz ])`
```

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 2]
```diff
@@ -483,6 +483,19 @@ in with passthru; stdenv.mkDerivation {
...
+     python3.pkgs.xyz', where xyz is the package you are trying to
```
Comment (RaitoBezarius) [inline pkgs/development/interpreters/python/cpython/default.nix:491]:
```suggestion
     where xyz is the package you are trying to
```

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 3]
```diff
@@ -483,6 +483,19 @@ in with passthru; stdenv.mkDerivation {
...
+     install.
+
+     If you wish to install a non-packaged Python package,
+     create a virtual environment using python3 -m venv path/to/venv.
+     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
+     sure you have python3-full installed.
```
Comment (RaitoBezarius) [inline pkgs/development/interpreters/python/cpython/default.nix:497]:
```suggestion
     sure you have python3-full installed, note that NixOS does not expose the classical library in your global environment.
     Therefore, many packages will not work out of the box, consider using a project like nix-ld: https://github.com/Mic92/nix-ld if this feature is important for you.
```

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 4]
```diff
@@ -24,6 +24,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - KDE Plasma has been updated to v5.27, see [the release notes](https://kde.org/announcements/plasma/5/5.27.0/) for what is changed.
 
+- Python implements PEP 668, providing better feedback to users that try to run `pip install` system-wide.
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:27]:
```suggestion
- Python implements [PEP 668](https://peps.python.org/pep-0668/), providing better feedback to users that try to run `pip install` system-wide.
```

**Comment by RaitoBezarius** [comment]
> > Now run `python3 -c "import numpy" `
> 
> the command completes with no output. checking the return code reveals success. It seems at least for some users, `pip install --user` is a valid and lightweight option, compared to heavy options like `venv`. do you have further evidence against `pip install --user`?

```
~ via ஃ via  v3.10.10 via ❄️  impure (shell) 
❯ pip install --user numpy           
Collecting numpy
  Downloading numpy-1.24.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 94.6 MB/s eta 0:00:00
Installing collected packages: numpy
  WARNING: The scripts f2py, f2py3 and f2py3.10 are installed in '/home/raito/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed numpy-1.24.3

[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python3.10 -m pip install --upgrade pip

~ via ஃ via  v3.10.10 via ❄️  impure (shell) took 3s 
❯ python                  
Python 3.10.10 (main, Feb  7 2023, 12:19:31) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'numpy'
```

On NixOS, so @mweinelt is right. `pip install --user` has no guarantee to work, furthermore, for the packages you mentioned, you are relying on wheels, which, by chance, are working at the moment.
But this is not a good enough guarantee.

**Comment by RaitoBezarius** [comment]
> How about something like this:
> 
> > `pip install` doesn't work with Nix because Nix is a purely functional package manager with a unique approach to package and environment management. Nix manages packages and dependencies in an isolated manner to provide reproducible builds and environments, whereas `pip` is a package installer for Python that works with more traditional environments.
> > To use Python with Nix, have a look at the online documentation: https://nixos.org/manual/nixpkgs/stable/#python.

(Ignore my suggestions.)

But I think it has to be related to **nixpkgs**, not Nix.
I would add a mention for NixOS also.

**Comment by FRidh** [comment]
> > Should nix profile install be mentioned here like in some other error messages?
> 
> Not while `nix profile` is still experimental
> 
> > Why isn't pip install --user mentioned in the error text?
> 
> `pip install --user` is also disabled by the PEP. I'm unsure why, but this is part of upstream's design. For some reason another tool called `pipx` is recommended for user installs on Debian.
> 
> > Thank you for looking into this. We should definitely have this file, however, I do not agree with the contents of the message.
> 
> +1 I wanted to show the capability and mostly took Debian's message and adapted it. The next step is to nail that message content down.
> 
> How about something like this:
> 
> > `pip install` doesn't work with Nix because Nix is a purely functional package manager with a unique approach to package and environment management. Nix manages packages and dependencies in an isolated manner to provide reproducible builds and environments, whereas `pip` is a package installer for Python that works with more traditional environments.
> > To use Python with Nix, have a look at the online documentation: https://nixos.org/manual/nixpkgs/stable/#python.

Sounds good to me! As @RaitoBezarius mentioned, it is indeed Python with/using Nixpkgs. I don't know how to exactly phrase it. I'd keep NixOS out of it here because it is also hard to express that succinct and making a change to this file is expensive. In the end users want to know how to go forward, and that is best described in the manual.

**Review by FRidh — APPROVED** [review]

**Comment by trofi** [comment]
Bisect says 3d2bdd0788d7fa05883cafc1f8c2ab29527ef4f1 `python: implement PEP 668 (#229166)` broke `python310Packages.flit` in `staging` as:

```
$ nix build --no-link -f. python310Packages.flit -L
...
python3.10-flit> =========================== short test summary info ============================
python3.10-flit> FAILED tests/test_install.py::InstallTests::test_install_data_dir - subprocess.CalledProcessError: Command '['/nix/store/02jdxqwwyx9sqqkm5zazfa...
python3.10-flit> FAILED tests/test_install.py::InstallTests::test_install_module_pep621 - subprocess.CalledProcessError: Command '['/nix/store/02jdxqwwyx9sqqkm5zazfa...
python3.10-flit> FAILED tests/test_install.py::InstallTests::test_symlink_data_dir - subprocess.CalledProcessError: Command '['/nix/store/02jdxqwwyx9sqqkm5zazfa...
python3.10-flit> FAILED tests/test_install.py::InstallTests::test_symlink_module_pep621 - subprocess.CalledProcessError: Command '['/nix/store/02jdxqwwyx9sqqkm5zazfa...
python3.10-flit> ============ 4 failed, 198 passed, 2 skipped, 1 deselected in 7.92s ============
```

**Comment by zimbatm** [comment]
Thanks, I'll look into it.

For Internet visitors that stumble onto this PR, here is a pretty good read about PEP 668, why it's also disabled for `pip install --user` and what options are available on typical distros: https://pythonspeed.com/articles/externally-managed-environment-pep-668/

**Pull request merged**